### ISSUE TITLE
ERROR: FPM initialization failed

### ISSUE THREAD

**Issue by ferremek** [laravel/valet#920]
Hi guys,

I have been trying to resolve this problem for a while already but it still comes back about once a week. I have Valet installed with Brew and it works all good but for one site, it just does not want to load every once in a while.

My php-fpm.log is full of the following error messages:
```
[10-Mar-2020 10:01:49] NOTICE: [pool valet] 'user' directive is ignored when FPM is not running as root
[10-Mar-2020 10:01:49] NOTICE: [pool valet] 'user' directive is ignored when FPM is not running as root
[10-Mar-2020 10:01:49] NOTICE: [pool valet] 'group' directive is ignored when FPM is not running as root
[10-Mar-2020 10:01:49] NOTICE: [pool valet] 'group' directive is ignored when FPM is not running as root
[10-Mar-2020 10:01:49] ERROR: Another FPM instance seems to already listen on /Users/ferre/.config/valet/valet.sock
[10-Mar-2020 10:01:49] ERROR: Another FPM instance seems to already listen on /Users/ferre/.config/valet/valet.sock
[10-Mar-2020 10:01:49] ERROR: FPM initialization failed
[10-Mar-2020 10:01:49] ERROR: FPM initialization failed
```

The nginx-error.log has the following errors:
```
2020/03/10 09:45:55 [alert] 308#0: *22804 setsockopt(TCP_NODELAY) failed (22: Invalid argument) while keepalive, client: 127.0.0.1, server: 127.0.0.1:80
2020/03/10 09:59:42 [alert] 308#0: *23101 setsockopt(TCP_NODELAY) failed (22: Invalid argument) while keepalive, client: 127.0.0.1, server: 127.0.0.1:80
2020/03/10 09:59:42 [alert] 307#0: *23115 setsockopt(TCP_NODELAY) failed (22: Invalid argument) while keepalive, client: 127.0.0.1, server: 127.0.0.1:80
2020/03/10 10:01:48 [alert] 301#0: *23133 setsockopt(TCP_NODELAY) failed (22: Invalid argument) while keepalive, client: 127.0.0.1, server: 127.0.0.1:80
```

I have the following Brew services installed:
```
Name    Status  User Plist
dnsmasq  started    root /Library/LaunchDaemons/homebrew.mxcl.dnsmasq.plist
mariadb    started    root /Library/LaunchDaemons/homebrew.mxcl.mariadb.plist
nginx        started    root /Library/LaunchDaemons/homebrew.mxcl.nginx.plist
php          stopped
php@7.3  started    root /Library/LaunchDaemons/homebrew.mxcl.php@7.3.plist
redis        started    root /Library/LaunchDaemons/homebrew.mxcl.redis.plist
```

For now I am stuck and don't know where to begin resolving this issue, I did install PH 7.3 and that resolved the issue for a little while but now it's back again.

Any ideas in what direction I need to think to resolve this?

Thanks a lot!

Cheers,
Ferre

**Comment by drbyte** [comment]
What's the output of the following?

```sh
sw_vers
valet --version
cat ~/.config/valet/config.json
cat ~/.composer/composer.json
ls -al /etc/sudoers.d/
brew config
brew services list
brew list --versions | grep -E '(php|nginx|dnsmasq|mariadb|mysql|mailhog|openssl)(@\d\..*)?\s'
php -v
which -a php
nginx -v
brew info nginx
brew info php
brew info php@7.3
sudo nginx -t
which -a php-fpm
sudo /usr/local/opt/php/sbin/php-fpm --test
sudo /usr/local/opt/php/sbin/php-fpm -y /usr/local/etc/php/7.3/php-fpm.conf --test
ls -al ~/Library/LaunchAgents
ls -al /Library/LaunchAgents
ls -al /Library/LaunchDaemons
```

I suspect your FPM startup errors about not running as root are caused by another actual PHP process running, either from another 7.3 FPM config file, or a different PHP version config file.
This seems to be confirmed by the "another FPM instance is running".
And it could be a result of having installed multiple PHP versions and not having old configs properly removed previously. The LaunchDaemon and LaunchAgent checks can help there.

More and more I'm seeing people having these issues when they fail to run `brew upgrade` and `composer global update` regularly, and when they manually start/stop brew services as their own user instead of using valet commands so they run as root.

**Comment by ferremek** [comment]
@drbyte Thanks for your response! The output for the commands as follows:

```
ferre ~ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.2
BuildVersion:	19C57
```

```
ferre ~ valet --version
Laravel Valet 2.8.1
```

```
ferre ~ cat ~/.config/valet/config.json
{
    "tld": "test",
    "paths": [
        "/Users/ferre/.config/valet/Sites",
        "/Users/ferre/Sites"
    ]
}
```

```
ferre ~ cat ~/.composer/composer.json
{
    "require": {
        "laravel/installer": "^2.1",
        "laravel/valet": "^2.8"
    }
}
```

```
ferre ~ ls -al /etc/sudoers.d/
total 16
drwxr-xr-x   4 root  wheel   128 15 jan 13:23 .
drwxr-xr-x  85 root  wheel  2720 10 mrt 10:05 ..
-rw-r--r--   1 root  wheel    80 15 jan 13:23 brew
-rw-r--r--   1 root  wheel    83 15 jan 13:23 valet
```

```
ferre ~ brew config
HOMEBREW_VERSION: 2.2.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 27fa87c94a6cf7be40fc8f8fc96bc7c387b7781e
Last commit: 3 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 966cffabec8db88724bc98edbcc52ab7c6bc4594
Core tap last commit: 13 days ago
HOMEBREW_PREFIX: /usr/local
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
Clang: 11.0 build 1100
Git: 2.21.0 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.2-x86_64
CLT: 11.3.0.0.1.1574140115
Xcode: 11.3.1
```

```
ferre ~ brew services list
Name    Status  User Plist
dnsmasq started root /Library/LaunchDaemons/homebrew.mxcl.dnsmasq.plist
mariadb started root /Library/LaunchDaemons/homebrew.mxcl.mariadb.plist
nginx   started root /Library/LaunchDaemons/homebrew.mxcl.nginx.plist
php     stopped
php@7.3 started root /Library/LaunchDaemons/homebrew.mxcl.php@7.3.plist
redis   started root /Library/LaunchDaemons/homebrew.mxcl.redis.plist
```

```
ferre ~ brew list --versions | grep -E '(php|nginx|dnsmasq|mariadb|mysql|mailhog|openssl)(@\d\..*)?\s'
curl-openssl 7.68.0
dnsmasq 2.80
mariadb 10.4.11
nginx 1.17.8
openssl 1.0.2s
openssl@1.1 1.1.1d
php 7.4.2
php@7.3 7.3.15
```

```
ferre ~ php -v
PHP 7.3.15 (cli) (built: Feb 20 2020 10:37:19) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.15, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.15, Copyright (c) 1999-2018, by Zend Technologies
```

```
ferre ~ which -a php
/usr/local/bin/php
/usr/bin/php
```

```
ferre ~ nginx -v
nginx version: nginx/1.17.8
```

```
ferre ~ brew info nginx
nginx: stable 1.17.8 (bottled), HEAD
HTTP(S) server and reverse proxy, and IMAP/POP3 proxy server
https://nginx.org/
/usr/local/Cellar/nginx/1.17.8 (25 files, 2.1MB) *
  Poured from bottle on 2020-01-30 at 11:23:45
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/nginx.rb
==> Dependencies
Required: openssl@1.1 ✔, pcre ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
Docroot is: /usr/local/var/www

The default port has been set in /usr/local/etc/nginx/nginx.conf to 8080 so that
nginx can run without sudo.

nginx will load all files in /usr/local/etc/nginx/servers/.

To have launchd start nginx now and restart at login:
  brew services start nginx
Or, if you don't want/need a background service you can just run:
  nginx
==> Analytics
install: 34,052 (30 days), 106,677 (90 days), 418,817 (365 days)
install-on-request: 32,877 (30 days), 102,622 (90 days), 400,148 (365 days)
build-error: 0 (30 days)
```

>  I removed some of the following output here because it's quite long!
```
ferre ~ brew info php
php: stable 7.4.3 (bottled)
General-purpose scripting language
https://www.php.net/
/usr/local/Cellar/php/7.4.2 (517 files, 76.0MB)
```

>  I removed some of the following output here because it's quite long!
```
ferre ~ brew info php@7.3
php@7.3: stable 7.3.15 (bottled) [keg-only]
General-purpose scripting language
https://www.php.net/
/usr/local/Cellar/php@7.3/7.3.15 (521 files, 77MB) *
```

```
ferre ~ sudo nginx -t
nginx: the configuration file /usr/local/etc/nginx/nginx.conf syntax is ok
nginx: configuration file /usr/local/etc/nginx/nginx.conf test is successful
```

```
ferre ~ which -a php-fpm
/usr/local/sbin/php-fpm
/usr/sbin/php-fpm
```

```
ferre ~ sudo /usr/local/opt/php/sbin/php-fpm --test
[10-Mar-2020 11:14:54] NOTICE: configuration file /usr/local/etc/php/7.4/php-fpm.conf test is successful
```

```
ferre ~ sudo /usr/local/opt/php/sbin/php-fpm -y /usr/local/etc/php/7.3/php-fpm.conf --test
[10-Mar-2020 11:15:10] NOTICE: configuration file /usr/local/etc/php/7.3/php-fpm.conf test is successful
```

```
ferre ~ ls -al ~/Library/LaunchAgents
total 104
drwxr-xr-x@ 15 ferre  staff   480  6 mrt 21:38 .
drwx------@ 81 ferre  staff  2592 13 feb 13:55 ..
-rw-r--r--   1 ferre  staff   687 17 aug  2019 com.adobe.GC.Invoker-1.0.plist
-rw-r--r--   1 ferre  staff   491 16 aug  2019 com.adobe.ccxprocess.plist
-rw-r--r--   1 ferre  staff   685 22 aug  2019 com.dropbox.DropboxMacUpdate.agent.plist
-rw-r--r--   1 root   staff  1502 16 aug  2019 com.expressvpn.ExpressVPN.agent.plist
-rw-r--r--@  1 ferre  staff   809  6 mrt 21:38 com.google.keystone.agent.plist
-rw-r--r--@  1 ferre  staff   915  6 mrt 21:38 com.google.keystone.xpcservice.plist
-rw-r--r--   1 ferre  staff   657 26 feb 10:21 homebrew.mxcl.dnsmasq.plist
-rw-r--r--   1 ferre  staff   541 26 feb 10:54 homebrew.mxcl.mariadb.plist
-rw-r--r--   1 ferre  staff   543 15 jan 13:48 homebrew.mxcl.mysql.plist
-rw-r--r--   1 ferre  staff   571 26 feb 10:21 homebrew.mxcl.nginx.plist
-rw-r--r--   1 ferre  staff   636 26 feb 10:21 homebrew.mxcl.php@7.3.plist
-rw-r--r--   1 ferre  staff   823 26 feb 10:21 homebrew.mxcl.redis.plist
-rw-r--r--@  1 ferre  staff   677 12 sep 18:34 org.virtualbox.vboxwebsrv.plist
```

```
ferre ~ ls -al /Library/LaunchAgents
total 48
drwxr-xr-x   8 root  wheel   256 13 feb 19:25 .
drwxr-xr-x  70 root  wheel  2240  6 jan 19:41 ..
-rw-r--r--   1 root  wheel   577 13 feb 19:25 com.adobe.ARMDCHelper.cc24aef4a1b90ed56a725c38014c95072f92651fb65e1bf9c8e43c37a23d420d.plist
-rw-r--r--   1 root  wheel   667 16 aug  2019 com.adobe.AdobeCreativeCloud.plist
-rw-r--r--   1 root  wheel   588 17 aug  2019 com.adobe.GC.AGM.plist
-rw-r--r--   1 root  wheel   612 17 aug  2019 com.adobe.GC.Invoker-1.0.plist
-rw-r--r--   1 root  wheel   668  7 nov 00:34 com.teamviewer.teamviewer.plist
-rw-r--r--   1 root  wheel   833  7 nov 00:34 com.teamviewer.teamviewer_desktop.plist
```

```
ferre ~ ls -al /Library/LaunchDaemons
total 88
drwxr-xr-x  14 root  wheel   448  2 mrt 18:28 .
drwxr-xr-x  70 root  wheel  2240  6 jan 19:41 ..
-rw-r--r--   1 root  wheel   474  7 feb 19:40 com.adobe.ARMDC.Communicator.plist
-rw-r--r--   1 root  wheel   486  7 feb 19:40 com.adobe.ARMDC.SMJobBlessHelper.plist
-rw-r--r--   1 root  wheel   564  5 okt 18:22 com.adobe.acc.installer.v2.plist
-rwxr-xr-x   1 root  wheel   642 23 okt 16:32 com.adobe.agsservice.plist
-rw-r--r--   1 root  wheel   544  7 nov 00:34 com.teamviewer.Helper.plist
-rw-r--r--   1 root  wheel   611  7 nov 00:34 com.teamviewer.teamviewer_service.plist
-rw-r--r--   1 root  admin   657  2 mrt 18:27 homebrew.mxcl.dnsmasq.plist
-rw-r--r--   1 root  admin   541 26 feb 11:02 homebrew.mxcl.mariadb.plist
-rw-r--r--   1 root  admin   571  2 mrt 18:28 homebrew.mxcl.nginx.plist
-rw-r--r--   1 root  admin   636  2 mrt 18:27 homebrew.mxcl.php@7.3.plist
-rw-r--r--   1 root  admin   823 26 feb 11:03 homebrew.mxcl.redis.plist
lrwxr-xr-x   1 root  wheel    76 12 sep 18:34 org.virtualbox.startup.plist -> ../Application Support/VirtualBox/LaunchDaemons/org.virtualbox.startup.plist
```

**Comment by drbyte** [comment]
Notice how you've got several homebrew LaunchAgents in your home directory that duplicate the ones in your system LaunchDaemons directory? 
The ones in your home directory run as "you", not as "root", and this is probably what's causing the problem you reported.

First run `valet stop`.

Next, `brew services cleanup` may be a good step.
If it doesn't give any feedback then I'd just also manually remove all those other agents:
`rm -f ~/Library/LaunchAgents/homebrew.mxcl.*.plist`

and `valet start`

**Comment by ferremek** [comment]
@drbyte Thanks again for your response!

I've stopped Valet and try to clean up Brew, it says there's nothing to be cleaned.

```
ferre ~ brew services cleanup
All user-space services OK, nothing cleaned...
```

After that I have removed all the agents from the LaunchAgents folder and restarted Valet.

```
ferre ~ valet start
Restarting dnsmasq...
Restarting php@7.3...
Restarting nginx...
Valet services have been started.
```

But when I check my php-fpm.log file I still see the same errors occur...
```
[10-Mar-2020 12:00:16] ERROR: FPM initialization failed
[10-Mar-2020 12:00:16] ERROR: FPM initialization failed
[10-Mar-2020 12:00:26] ERROR: Another FPM instance seems to already listen on /Users/ferre/.config/valet/valet.sock
[10-Mar-2020 12:00:26] ERROR: Another FPM instance seems to already listen on /Users/ferre/.config/valet/valet.sock
[10-Mar-2020 12:00:26] ERROR: FPM initialization failed
[10-Mar-2020 12:00:26] ERROR: FPM initialization failed
```

**Comment by drbyte** [comment]
I think a reboot will help, now that those LaunchAgents have been removed, which in theory should now be preventing the extra fpm instance from running.

After that, let's look at your `/usr/local/etc/php/7.3/php-fpm.d/*.conf` files ... and the 7.4 equivalents. Ideally there should only be ONE `*.conf` file in each of those directories (specifically, `valet-fpm.conf`).
But if you'd copied/cloned any of the valet FPM config files then it could be that you've created a conflict by trying to just keep backups.

**Comment by ferremek** [comment]
@drbyte That's it, it's resolved! Thank you so much for your help!

One more question, you said the following:
> More and more I'm seeing people having these issues when they fail to run `brew upgrade` and `composer global update` regularly, and when they manually start/stop brew services as their own user instead of using valet commands so they run as root.

How often would you recommend to do these updates and upgrade commands?

**Comment by drbyte** [comment]
@ferremek I'm glad it's resolved.

To prevent those "user-instances" (instead of "root" instances), avoid running `brew start` commands without `sudo` ... or just use Valet to start/stop the services related to Valet.

As for upgrades, I'm in the habit of running `brew upgrade` weekly, `brew cleanup` monthly, 
and `composer global update` every 2-4 weeks.

**Comment by drbyte** [comment]
@ferremek if this is resolved for you, please help us keep the repo clean by clicking Close below.

**Comment by ferremek** [comment]
@drbyte Thanks again for your advice!

**Comment by ihorvorotnov** [comment]
Just a quick note - noticed the same issue today, `php-fpm` was running as root (launched by valet), but at the same time brew tried to launch it for some reason too and the service was reporting an error. I just ran `brew services stop php` and the problems was solved - launch agent removed, service stopped, errors gone. The other instance managed by valet (and launched as root) was left intact.

**Comment by ArnoHolo** [comment]
Just a quick note - noticed the same issue today.

It's not recommend at all to run brew as root !

I managed to fix this with by removing the valet socket : `valet stop && rm ~/.config/valet/valet.sock && valet start`

**Comment by KhurshidYakubov** [comment]
I also encountered such a problem because I have two versions of PHP(7.3 and 8.3) on my machine. 
I fixed this error just by stopping my PHP(8.3) from brew services. Because it was conflicting with my valet fpm:
`brew services stop php`

P.S. There is no need to define php version on command because 'php' in brew services means the latest version of it. It was exactly 8.3 at that time. If needed you can stop another version by defining it(ex. _brew services stop php@7.3_)
